# Exploring Hacker News Posts

## Introduction
[Hacker News](https://news.ycombinator.com) is a platform initiated by the startup incubator Y Combinator, where user submitted stories, referred to as posts, receive votes and comments. The site is highly popular within technology and startup communities, and posts reaching the top listings can attract hundreds of thousands of visitors. This project explores two specific categories of posts: `'Ask HN'` and `'Show HN'`. `'Ask HN'` posts are submitted to pose a specific question to the Hacker News community. Conversely, `'Show HN'` posts are used to share a project, product, or general item of interest with the community.

## The Goal
The primary objective of this analysis is to compare these two post categories to determine the following:
Whether `'Ask HN'` or `'Show HN'` posts receive more comments on average.
Whether posts created at specific times receive more comments on average.

## Data Source
The [dataset](https://dq-content.s3.amazonaws.com/356/hacker_news.csv) utilised for this analysis contains approximately 20,000 rows. It was derived from an [original dataset](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts) of nearly 300,000 rows by filtering out submissions with zero comments and subsequently taking a random sample from the remaining entries. 

The dataset comprises the following columns:
<br/>`id`: The unique identifier from Hacker News for the post.
<br/>`title`: The title of the post.
<br/>`url`: The URL that the post links to, if applicable.
<br/>`num_points`: The number of points the post acquired, calculated as the total number of upvotes less the total number of downvotes.
<br/>`num_comments`: The number of comments on the post.
<br/>`author`: The username of the person who submitted the post.
<br/>`created_at`: The date and time of the post submission.

## Read in the Data and Remove the Header

In [42]:
import csv 
with open("Dataqust project - Hackers/hacker_news.csv") as file:
    hn = list(csv.reader(file))

In [43]:
header = hn[0]
hn = hn[1:]

print(header)
print('\n')
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Extracting Ask HN and Show HN Posts

In [44]:
ask_posts = []
show_posts = []
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(post)
    elif title.lower().startswith("show hn"):
        show_posts.append(post)
    else:
        other_posts.append(post)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))


1744
1162
17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [45]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comment = 0

for post in ask_posts:
    num_comment = int(post[4])
    total_ask_comment += num_comment

avg_ask_comment = total_ask_comment / len(ask_posts)
print(avg_ask_comment)


14.038417431192661


In [46]:
# Calculate the average number of comments `Show HN` posts receive.
total_show_comment = 0

for post in show_posts:
    num_comment = int(post[4])
    total_show_comment += num_comment

avg_show_comment = total_show_comment / len(show_posts)
print(avg_show_comment)

10.31669535283993


On average, 'Ask HN' posts in the sample receive approximately 14 comments, whereas 'Show HN' posts receive approximately 10. Since 'Ask HN' posts are more likely to receive comments, the remaining analysis will focus exclusively on this category.

## Finding the Amount of Ask Posts and Comments by Hour Created



In [47]:
import datetime as dt
result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])])

count_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date,"%m/%d/%Y %H:%M")
    hour = time.strftime("%H")
    if hour in count_by_hour:
        comments_by_hour[hour] += comment
        count_by_hour[hour] += 1
    else:
        comments_by_hour[hour] = comment
        count_by_hour[hour] = 1

comments_by_hour


{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [48]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / count_by_hour[hour]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

## Sorting and Printing Values from a List of Lists

In [49]:
# Invert the positions of the hour and the average comment value.
# Python sorts lists of lists by the first element of each sublist by default.
# Swapping the elements places the numerical average first, ensuring the 
# subsequent sorting is based on comment volume rather than chronological hours.
swap_avg_by_hour = []
for hour in avg_by_hour:
    swap_avg_by_hour.append([hour[1], hour[0]])

print(swap_avg_by_hour)


[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [50]:
sort_swap = sorted(swap_avg_by_hour, reverse = True)
print(sort_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


## Finding Top 5 Hours for Ask Posts Comments

In [51]:
print("Top 5 Hour for Ask Posts Comments:")

for avg, hr in sort_swap[:5]:
    print(f"{dt.datetime.strptime(hr, "%H").strftime("%H:%M")} : {avg:2f} average comments per post")
    
    

Top 5 Hour for Ask Posts Comments:
15:00 : 38.594828 average comments per post
02:00 : 23.810345 average comments per post
20:00 : 21.525000 average comments per post
16:00 : 16.796296 average comments per post
21:00 : 16.009174 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Conclusion

This project analysed a downsampled dataset of Hacker News submissions to determine whether 'Ask HN' or 'Show HN' posts receive more engagement, and to identify if the time of creation influences comment volume. 

The analysis revealed that `'Ask HN'` posts attract a higher volume of engagement, receiving an average of approximately 14.04 comments per post, compared to the 10.32 average comments received by `'Show HN'` posts. 

Furthermore, the hour at which a post is created heavily influences community interaction. For `'Ask HN'` posts, the **top five** hours for attracting comments are:

* **15:00 (3:00 PM EST)**: 38.59 average comments per post
* **02:00 (2:00 AM EST)**: 23.81 average comments per post
* **20:00 (8:00 PM EST)**: 21.53 average comments per post
* **16:00 (4:00 PM EST)**: 16.80 average comments per post
* **21:00 (9:00 PM EST)**: 16.01 average comments per post

The data indicates that 15:00 is the most effective time to submit an `'Ask HN'` post, yielding a significant increase in interaction compared to all other hours. Based on these findings, to maximise community engagement, it is recommended to categorise submissions as `'Ask HN'` and publish them during the mid-afternoon window, specifically around 15:00 EST.